ocsgw/src/main/java/org/ostelco/ocsgw/datasource/protobuf/PubSubDataSource.kt

```diff
@@ -0,0 +1,130 @@
+
+    private var publisher: Publisher
+
+    init {
+
+        // init publisher
+        val topicName = ProjectTopicName.of(projectId, topicId)
+        val strSocketAddress = System.getenv("PUBSUB_EMULATOR_HOST")
+        publisher = if (!strSocketAddress.isNullOrEmpty()) {
+            val channel = ManagedChannelBuilder.forTarget(strSocketAddress).usePlaintext().build()
+            // Create a publisher instance with default settings bound to the topic
+            val channelProvider = FixedTransportChannelProvider.create(GrpcTransportChannel.create(channel))
+            Publisher.newBuilder(topicName)
+                    .setChannelProvider(channelProvider)
+                    .setCredentialsProvider(NoCredentialsProvider())
+                    .build()
+        } else {
+            Publisher.newBuilder(topicName).build()
+        }
+
+        // Instantiate an asynchronous message receiver
+        setupPubSubSubscriber(projectId, ccrSubscriptionId) { message, consumer ->
+            // handle incoming message, then ack/nack the received message
+            asyncDataSource.handleProtobufCcrAnswer(
+                    CreditControlAnswerInfo.parseFrom(message))
+            consumer.ack()
+        }
+
+        setupPubSubSubscriber(projectId, activateSubscriptionId) { message, consumer ->
+            // handle incoming message, then ack/nack the received message
+            asyncDataSource.handleProtobufActivateResponse(
+                    ActivateResponse.parseFrom(message))
+            consumer.ack()
+        }
+    }
+
+    override fun init() {
+
+    }
+
+    override fun handleRequest(context: CreditControlContext) {
+
+        val creditControlRequestInfo = asyncDataSource.handleRequest(context)
+
+        if (creditControlRequestInfo != null) {
+            val pubsubMessage = PubsubMessage.newBuilder()
+                    .setData(creditControlRequestInfo.toByteString())
+                    .build()
+
+            //schedule a message to be published, messages are automatically batched
+            val future = publisher.publish(pubsubMessage)
+
+            // add an asynchronous callback to handle success / failure
+            ApiFutures.addCallback(future, object : ApiFutureCallback<String> {
+
+                override fun onFailure(throwable: Throwable) {
+                    if (throwable is ApiException) {
+                        // details on the API exception
+                        logger.warn("Status code: {}", throwable.statusCode.code)
+                        logger.warn("Retrying: {}", throwable.isRetryable)
+                    }
+                    logger.warn("Error publishing active users list")
+                }
+
+                override fun onSuccess(messageId: String) {
+                    // Once published, returns server-assigned message ids (unique within the topic)
+                    logger.debug(messageId)
```

Comment:
anyway to know which message succeeded if this is a server assigned message id ?

Reply:
This might not be possible.

Reply:
ok, maybe just remove log line, as it has no context.

Reply:
There is no action to be taken for successful submission.  If there was, that would be handled there.  Better to remove it.

Reply:
👍